### DISCUSSION TITLE
MauiAuth Login B2C on Windows?

### DISCUSSION THREAD

**Discussion by plettb** [VladislavAntonyuk/MauiSamples#436] [Q&A] [answered]
Let me begin by saying that I REALLY appreciate your work to provide samples for Maui.  Developing on Maui is, at least for me, very tedious because it seems I encounter a new problem at every turn, and there just isn't enough help available.  It seems you may be the only one actually responsive to Maui issues, so THANK-YOU!!!

Today I have another issue:  I'm trying to get your example "MauiAuth" to allow for B2C login on Windows.  I've made the necessary changes to "Auth\AuthServices\Constants.cs", but when I attempt to login B2C on Windows, a browser opens up and navigates to `https://localhost:7157/signin-oidc?error=redirect_uri_mismatch&error_description=AADB2C90006%3a+The+redirect+URI+%27http%3a%2f%2flocalhost%3a60114%27+provided+in+the+request+is+not+registered+for+the+client+id+%27...`.

I assume this is a configuration issue, but I don't even know where to look.  Any help would be appreciated.

Thanks!

**Comment by VladislavAntonyuk** [comment]
Hey, double check configuration in Azure portal. According to the issue you haven’t registered localhost redirect url for your client.
Try my YouTube guide for it: https://youtu.be/sTPWF2O456U?si=6rfO_JdrITJzluoV

**Reply by plettb** [reply]
I feel like I got a little closer by hardcoding a redirect URI (http://localhost:53143) and adding that into the Azure configuration, but now I get "Clients must send a client_secret when redeeming a confidential grant."

**Reply by VladislavAntonyuk** [reply]
You can try the VladislavAntonyuk/WorldExplorer app. It should work (at least it worked last month). 
Just localhost should be enough. Also make sure you set Client to work with mobile and desktop (additional checkbox in Azure configuration page)

**Reply by plettb** [reply]
The WorldExplorer app looks like it _might_ be working, but regardless of what "WindowsMaps" key I use, when I try to run it I get ".NET MAUI Maps is currently not implemented for Windows". I commented out the maps in the DI, and then the login seems to work (at least I get further), but then I get "Handler not found for view Microsoft.Maui.Controls.Maps.Map" and then it crashes.  The reason I think the login is working, though, is because the next time running the app, I'm logged in.

On an unrelated note:
- The second time in I appear to be logged in, but "Explorer" just keeps "Loading...". It's quite possible this is related to commenting out the maps.
- Whereas I know "Logout" works (I'm logged out the next time I run), there are no visible changes.  Again, this may be due to the other errors.

Thanks again for your help!

**Reply by VladislavAntonyuk** [reply]
If you redirected to the map, then you have signed in. You can take auth service from the app (actually it should be the same as in MauiSamples).
Although, did you use my appsettings configuration for the auth or replaced with your own?
Try put my appsettings in your app. If you successfully sign in then the issue might be in your azure portal configuration

**Reply by plettb** [reply]
I have not been successful putting my settings into WorldExplorer.  It's unclear to me where you're getting the "AuthorizeUrl" and (especially) the "AuthoritySignIn".  I tried modifying them as follows, but that doesn't work:
```
"AuthorizeUrl": "https://{myappname}.b2clogin.com/{myappname}.onmicrosoft.com/oauth2/v2.0/authorize",
"AuthoritySignIn": "https://{myappname}.b2clogin.com/tfp/{myappname}.onmicrosoft.com/{mysigninpolicy}",
"SignInPolicy": "{mysigninpolicy}"
```
With those settings I get the redirect error (in the original message).

Assuming I have those correct, I'm guessing it must be the Azure B2C configuration, which I have configured as follows:
![image](https://github.com/user-attachments/assets/ea175786-a574-4fb1-b47c-f5cb10946cfe)
(Of course I've also tried to add the redirect URI that's in the error message, but it seems the port changes each time it's called.  Since your app doesn't appear to use a redirect URI, I assume the redirect URIs don't need to be configured.)

Given that WorldExplorer works so well, it all feels tantalizingly close, but....

Any ideas?

**Reply by VladislavAntonyuk** [reply]
my redirect URIs:
![image](https://github.com/user-attachments/assets/864cb362-d3fe-4f37-af90-8388b499f109)

**Reply by plettb** [reply, marked as answer]
Well, this is one of those weird ones where I'll likely never know what was actually wrong.  I finally created an entirely new B2C tenant from scratch and everything worked like a charm. :)

Thanks for your help, and sorry for bothering you.

**Reply by VladislavAntonyuk** [reply]
Glad to know that everything worked out for you.
Also I have updated WorldExplorer, maps should not bother you.
